File: mix.scala
object Color extends Enumeration {
  val Blue, Red, Magenta, Green, Cyan, Yellow, White = Value
}

object Main {
  import Color._

  def mix(a:Color.Value, b:Color.Value): Color.Value = {
    (a, b) match {
      case (Blue, Red)       => Magenta
      case (Blue, Magenta)   => Magenta
      case (Blue, Green)     => Cyan
      case (Blue, Cyan)      => Cyan
      case (Blue, Yellow)    => White
      case (Red, Magenta)    => Magenta
      case (Red, Green)      => Yellow
      case (Red, Cyan)       => White
      case (Red, Yellow)     => Yellow
      case (Magenta, Green)  => White
      case (Magenta, Cyan)   => White
      case (Magenta, Yellow) => White
      case (Green, Cyan)     => Cyan
      case (Green, Yellow)   => Yellow
      case (Cyan, Yellow)    => White
      case (White, _)        => White
      case (a, b) if a == b  => a
      case (a, b)            => mix(b, a)
    }
  }

  def main(args: Array[String]): Unit = {
    println(mix(Red, Red))
    println(mix(Red, Blue))
    println(mix(Blue, mix(Blue, Green)))
  }
}

File: mix.hs
data Color = Blue | Red | Magenta | Green | Cyan | Yellow | White
    deriving Show

mix Blue    Red     = Magenta
mix Blue    Magenta = Magenta
mix Blue    Green   = Cyan
mix Blue    Cyan    = Cyan
mix Blue    Yellow  = White
mix Red     Magenta = Magenta
mix Red     Green   = Yellow
mix Red     Cyan    = White
mix Red     Yellow  = Yellow
mix Magenta Green   = White
mix Magenta Cyan    = White
mix Magenta Yellow  = White
mix Green   Cyan    = Cyan
mix Green   Yellow  = Yellow
mix Cyan    Yellow  = White
mix White   _       = White
mix c1      c2      = mix c2 c1

main = do
    print $ mix Red Blue
    print $ mix Red $ mix Blue Green
